test/Stripe.hs

```diff
@@ -218,3 +249,581 @@ chargeTests =
+
+webhookTests :: TestTree
+webhookTests =
+  testGroup "The Stripe charge web hook"
+  [ testCase "If the signature is missing then the response is Bad Request" $ do
+      db <- memory
+
+      let
+        theSRequest = SRequest jsonRequest checkoutSessionCompleted
+        app = makeApp db
+
+      response <- (flip runSession) app $ srequest theSRequest
+      assertEqual "The body reflects the error" (Just $ Failure "missing signature") (decode . simpleBody $ response)
+      assertEqual "The response is 400" status400 (simpleStatus response)
+      assertNotRedeemable db voucher fingerprint
+
+  , testCase "If the signature is misformatted then the response is Bad Request" $ do
+      db <- memory
+      let
+        app = makeApp db
+        theRequest = signedRequest "Do you like my signature?"
+        theSRequest = SRequest theRequest checkoutSessionCompleted
+
+      response <- (flip runSession) app $ srequest theSRequest
+      assertEqual "The body reflects the error" (Just $ Failure "malformed signature") (decode . simpleBody $ response)
+      assertEqual "The response is 400" status400 (simpleStatus response)
+      assertNotRedeemable db voucher fingerprint
+
+  , testCase "If the signature is incorrect then no attempt is made to parse the request body and the response is Bad Request" $ do
+      db <- memory
+      let
+        app = makeApp db
+        theRequest = signedRequest $ stripeSignature (WebhookSecretKey "key") timestamp "Some other body"
+        theSRequest = SRequest theRequest checkoutSessionCompleted
+
+      response <- (flip runSession) app $ srequest theSRequest
+      assertEqual "The body reflects the error" (Just $ Failure "invalid signature") (decode . simpleBody $ response)
+      assertEqual "The response is 400" status400 (simpleStatus response)
+      assertNotRedeemable db voucher fingerprint
+
+  , testCase "If the signature is correct and the body is not JSON then the response is Bad Request" $ do
+      db <- memory
+      let
+        nonJSONBody = "Some other body"
+        app = makeApp db
+        theRequest = signedRequest $ stripeSignature webhookSecret timestamp nonJSONBody
+        theSRequest = SRequest theRequest (LBS.fromStrict nonJSONBody)
+
+      response <- (flip runSession) app $ srequest theSRequest
+
+      -- It should fail but we don't really care what the message is.
+      let (Just (Failure _)) = decode . simpleBody $ response
+      assertEqual "The response is 400" status400 (simpleStatus response)
+
+  , testCase "If the signature is correct then the response is OK" $ do
+      db <- assertOkResponse checkoutSessionCompleted
+      -- It has been paid so we should be allowed to redeem it.
+      assertRedeemable db voucher fingerprint
+
+  , testCase "The response to a charge.succeeded is OK" $ do
+      db <- assertOkResponse chargeSucceeded
+      -- It is only redeemable after checkout.session.completed.
+      assertNotRedeemable db voucher fingerprint
```

Comment:
My conception of "charge succeeded" probably differs from Stripe's - Why should why not redeem a voucher when the charge has succeeded?

Reply:
Tried to clarify in 4684edf